examples/boid_flockers/boid_flockers/model.py

```diff
@@ -8,7 +8,107 @@
+        - Alignment: try to fly in the same direction as the neighbors.
+
+    Boids have a vision that defines the radius in which they look for their
+    neighbors to flock with. Their speed (a scalar) and direction (a vector)
+    define their movement. Separation is their desired minimum distance from
+    any other Boid.
+    """
+
+    def __init__(
+        self,
+        unique_id,
+        model,
+        pos,
+        speed,
+        direction,
+        vision,
+        separation,
+        cohere=0.025,
+        separate=0.25,
+        match=0.04,
+    ):
+        """
+        Create a new Boid flocker agent.
+
+        Args:
+            unique_id: Unique agent identifyer.
+            pos: Starting position
+            speed: Distance to move per step.
+            direction: numpy vector for the Boid's direction of movement.
+            vision: Radius to look around for nearby Boids.
+            separation: Minimum distance to maintain from other Boids.
+            cohere: the relative importance of matching neighbors' positions
+            separate: the relative importance of avoiding close neighbors
+            match: the relative importance of matching neighbors' headings
+        """
+        super().__init__(unique_id, model)
+        self.pos = np.array(pos)
+        self.speed = speed
+        self.direction = direction
+        self.vision = vision
+        self.separation = separation
+        self.cohere_factor = cohere
+        self.separate_factor = separate
+        self.match_factor = match
+
+    def cohere(self, neighbors):
+        """
+        Return the vector toward the center of mass of the local neighbors.
+        """
+        cohere = np.zeros(2)
+        if neighbors:
+            for neighbor in neighbors:
+                cohere += self.model.space.get_heading(self.pos, neighbor.pos)
+            cohere /= len(neighbors)
+        return cohere
+
+    def separate(self, neighbors):
+        """
+        Return a vector away from any neighbors closer than separation dist.
+        """
+        me = self.pos
+        them = (n.pos for n in neighbors)
+        separation_vector = np.zeros(2)
+        for other in them:
+            if self.model.space.get_distance(me, other) < self.separation:
+                separation_vector -= self.model.space.get_heading(me, other)
+        return separation_vector
+
+    def match_heading(self, neighbors):
+        """
+        Return a vector of the neighbors' average heading.
+        """
+        match_vector = np.zeros(2)
+        if neighbors:
+            for neighbor in neighbors:
+                match_vector += neighbor.direction
+            match_vector /= len(neighbors)
+        return match_vector
+
+    def step(self):
+        """
+        Get the Boid's neighbors, compute the new vector, and move accordingly.
+        """
+
+        neighbors = self.model.space.get_neighbors(self.pos, self.vision, False)
```

Comment:
If we make this self.neighbors we can also use if for the visualisation right?

Reply:
Yes that would work fine, but it could take too much memory with many agents. If the only need is to change colour, I think it would be more convenient to make a self.colour variable instead.

Reply:
We could save the neighbours of the previous step.

I think in many systems that compute will be more of an bottleneck than memory

Reply:
I have made the changes as suggested

Reply:
> The model can now smoothly handle around 600 agents after which it slows down. After 1000 it is too slow and not practical

That's useful info to take a note on.

Reply:
First it was around 200 right? If so, that's a great improvement! Thanks!